src/bootman/bootman.c

```diff
@@ -446,6 +461,188 @@ bool boot_manager_set_uname(BootManager *self, const char *uname)
+                                return false;
+                        }
+                }
+        }
+        return true;
+}
+
+bool boot_manager_remove_initrd_freestanding(BootManager * self)
+{
+        autofree(char) *base_path = NULL;
+        autofree(char) *initrd_efi_path = NULL;
+        autofree(DIR) *initrd_dir = NULL;
+        struct dirent *ent = NULL;
+        bool is_uefi = ((self->bootloader->get_capabilities(self) & BOOTLOADER_CAP_UEFI) ==
+                        BOOTLOADER_CAP_UEFI);
+        const char *efi_boot_dir =
+            is_uefi ? self->bootloader->get_kernel_destination(self) : NULL;
+        if (!self || !self->initrd_freestanding_dir) {
+                return false;
+        }
+        /* if it's UEFI, then bootloader->get_kernel_dst() must return a value. */
+        if (is_uefi && !efi_boot_dir) {
+                return false;
+        }
+
+        base_path = boot_manager_get_boot_dir((BootManager *)self);
+
+        initrd_efi_path = string_printf("%s/%s",
+                                        base_path,
+                                        (is_uefi ? efi_boot_dir : ""));
+
+        initrd_dir = opendir(initrd_efi_path);
+        if (!initrd_dir) {
+                LOG_ERROR("Error opening %s: %s", initrd_efi_path, strerror(errno));
+                return false;
+        }
+
+        while ((ent = readdir(initrd_dir)) != NULL) {
+                autofree(char) *initrd_target = NULL;
+
+                if (strstr(ent->d_name, "freestanding-") != ent->d_name) {
+                        continue;
+                }
+
+                if (!nc_hashmap_get(self->initrd_freestanding, ent->d_name)) {
+                        initrd_target = string_printf("%s/%s",
+                                                      initrd_efi_path,
+                                                      ent->d_name);
+                        /* Remove old initrd */
+                        if (nc_file_exists(initrd_target)) {
+                                if (unlink(initrd_target) < 0) {
+                                        LOG_ERROR("Failed to remove legacy-path UEFI initrd %s: %s",
+                                                  initrd_target,
+                                                  strerror(errno));
+                                        return false;
+                                }
+                        }
+                }
+        }
+        return true;
+}
+
+void boot_manager_initrd_iterator_init(const BootManager *manager, NcHashmapIter *iter)
+{
+        if (!iter) {
+                return;
+        }
+        nc_hashmap_iter_init(manager->initrd_freestanding, iter);
+}
+
+bool boot_manager_enumerate_initrd_freestanding(NcHashmapIter *iter, const char *name)
+{
+        void *key = NULL;
+        void *val = NULL;
+
+        if (!iter || !name) {
+                return false;
+        }
+
+        if (nc_hashmap_iter_next(iter, &key, &val)) {
```

Comment:
Is there any specific order in which the initrds are returned by the iterator? Or it's just based on they hash value?

Reply:
order doesn't matter every file in the initrds will be allocated in the same space of memory